ENH: Adds support for central location to store parameters and more!

rstoneback opened the pull request:
# Description

Addresses #123, #183, #309, #496, #542, and #550 so far.

Adds initial support for a single place to store all pysat related parameters that is also thread safe. This feature impacts a variety of areas. I've added some new parameters (`warn_empty_file_list` to support #550) and coupled these into default Instrument instantiation, if a user doesn't provide input. Simple bug fixes are included as well.

I've also taken a pass at improving the Files class. I've worked on the docstrings a bit, added a number of attributes to the 
top-level docstring. I didn't touch the manual_org as it is handled elsewhere. I was able to remove the weak reference back to the pysat.Instrument object from within files. I copy over relevant parameters instead. This did require a small tweak to __repr__.

I added some stub code for supporting a `list_files` type of method that generates filenames on demand, rather than using a stored list. This is intended for pysatMissions. See caveats below.

 I also internally moved some code around a bit to filter internal file lists before they are stored, rather than upon every load.
I tried to make the warnings and logging statements more useful.

Examples for the user visible changes listed are included below.

The actual directory structure on disk can now be updated via command.


## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality
      to not work as expected)
- This change requires a documentation update

# How Has This Been Tested?

Unit tests have been expanded.

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

# Examples:

```
# pysat info is now stored in .pysat/pysat_settings.json
In [1]: import pysat                                                                                                                                                                                                        

In [2]: print(pysat.params)                                                                                                                                                                                                 
pysat Parameters object
----------------------
Tracking 7 pysat settings
Tracking 1 settings (non-default)
Tracking 0 user values

Standard parameters:
clean_level : clean
directory_format : {platform}/{name}/{tag}/{inst_id}
ignore_empty_files : False
file_timeout : 10
update_files : True
user_modules : {}
warn_empty_file_list : False

Standard parameters (no defaults):
data_dirs : ['/Users/russellstoneback/ScienceData/']

User parameters:
```
Please note the updated data directory structure.

To set a new parameter, simply assign to the params dict.
```
In [3]: pysat.params['update_files'] = False                                                                                                                                                                                

In [6]: pysat.params['data_dirs'] = '~/ScienceData'                                                                                                                                                                         

In [7]: print(pysat.params)                                                                                                                                                                                                 
pysat Parameters object
----------------------
Tracking 7 pysat settings
Tracking 1 settings (non-default)
Tracking 0 user values

Standard parameters:
clean_level : clean
directory_format : {platform}/{name}/{tag}/{inst_id}
ignore_empty_files : False
file_timeout : 10
update_files : False
user_modules : {}
warn_empty_file_list : False

Standard parameters (no defaults):
data_dirs : ['/Users/russellstoneback/ScienceData']

User parameters:
```
Currently all settings are stored to disk whenever one is updated.

`pysat.data_dir` has been removed. I've tried the getter/setter property trick to get `pysat.data_dir` to transparently resolve to `pysat.params['data_dirs'][0]` and raise a DeprecationWarning. Unfortunately, it isn't working.

Meta is now immutable at the start, by default.
```
In [8]: inst = pysat.Instrument(platform='pysat', name='testing')                                                                                                                                                           

In [9]: inst.meta.hi = 'hi!'                                                                                                                                                                                                
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-180098cc335d> in <module>
----> 1 inst.meta.hi = 'hi!'

~/Code/pysat/pysat/_meta.py in __setattr__(self, name, value)
    310                                      "since the Meta object attributes are",
    311                                      "set to immutable.")).format(val=value)
--> 312                     raise AttributeError(estr)
    313         else:
    314             super(Meta, self).__setattr__(name, value)

AttributeError: Cannot set attribute hi to hi! since the Meta object attributes are set to immutable.
```

The .pysat directory structure has also been updated.
```
In [10]: cd ~/.pysat                                                                                                                                                                                                        
/Users/russellstoneback/.pysat

In [11]: ls                                                                                                                                                                                                                 
instruments/         pysat_settings.json

In [12]: cd instruments/                                                                                                                                                                                                    
/Users/russellstoneback/.pysat/instruments

In [13]: ls                                                                                                                                                                                                                 
archive/                                   pysat_testing___stored_file_info.txt       pysat_testing__hello_stored_file_info.txt
```

We can now also transition from one directory structure to another. Currently it is set up for printing to terminal rather than logging. Also, by default it is only a test run.
```
In [4]: pysat.utils.files.update_data_directory_structure('{platform}', full_breakdown=True)                                                                                                                                
Working Instrument: sw dst  
Current path :  /Users/russellstoneback/ScienceData/sw/dst/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
5 files located.

Will move: /Users/russellstoneback/ScienceData/sw/dst/dst2001.txt
       to: /Users/russellstoneback/ScienceData/sw/dst2001.txt
Will move: /Users/russellstoneback/ScienceData/sw/dst/dst2002.txt
       to: /Users/russellstoneback/ScienceData/sw/dst2002.txt
Will move: /Users/russellstoneback/ScienceData/sw/dst/dst2003.txt
       to: /Users/russellstoneback/ScienceData/sw/dst2003.txt
Will move: /Users/russellstoneback/ScienceData/sw/dst/dst2004.txt
       to: /Users/russellstoneback/ScienceData/sw/dst2004.txt
Will move: /Users/russellstoneback/ScienceData/sw/dst/dst2005.txt
       to: /Users/russellstoneback/ScienceData/sw/dst2005.txt

Working Instrument: sw f107  
Current path :  /Users/russellstoneback/ScienceData/sw/f107/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
11 files located.

Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2009-01.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2009-01.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2009-02.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2009-02.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2009-03.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2009-03.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-01.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-01.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-02.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-02.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-03.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-03.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-04.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-04.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-05.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-05.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-06.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-06.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-07.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-07.txt
Will move: /Users/russellstoneback/ScienceData/sw/f107/f107_monthly_2015-08.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_monthly_2015-08.txt

Working Instrument: sw f107 all 
Current path :  /Users/russellstoneback/ScienceData/sw/f107/all/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
1 files located.

Will move: /Users/russellstoneback/ScienceData/sw/f107/all/f107_1947_to_2021-01-07.txt
       to: /Users/russellstoneback/ScienceData/sw/f107_1947_to_2021-01-07.txt

Working Instrument: sw f107 prelim 
Current path :  /Users/russellstoneback/ScienceData/sw/f107/prelim/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw f107 daily 
Current path :  /Users/russellstoneback/ScienceData/sw/f107/daily/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw f107 forecast 
Current path :  /Users/russellstoneback/ScienceData/sw/f107/forecast/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw f107 45day 
Current path :  /Users/russellstoneback/ScienceData/sw/f107/45day/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw kp  
Current path :  /Users/russellstoneback/ScienceData/sw/kp/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
12 files located.

Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0101.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0101.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0102.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0102.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0103.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0103.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0104.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0104.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0105.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0105.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0106.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0106.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0107.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0107.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0108.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0108.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0109.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0109.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0110.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0110.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0111.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0111.tab
Will move: /Users/russellstoneback/ScienceData/sw/kp/kp0112.tab
       to: /Users/russellstoneback/ScienceData/sw/kp0112.tab

Working Instrument: sw kp forecast 
Current path :  /Users/russellstoneback/ScienceData/sw/kp/forecast/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw kp recent 
Current path :  /Users/russellstoneback/ScienceData/sw/kp/recent/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace realtime mag
Current path :  /Users/russellstoneback/ScienceData/sw/ace/realtime/mag/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace historic mag
Current path :  /Users/russellstoneback/ScienceData/sw/ace/historic/mag/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace realtime swepam
Current path :  /Users/russellstoneback/ScienceData/sw/ace/realtime/swepam/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace historic swepam
Current path :  /Users/russellstoneback/ScienceData/sw/ace/historic/swepam/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace realtime epam
Current path :  /Users/russellstoneback/ScienceData/sw/ace/realtime/epam/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace historic epam
Current path :  /Users/russellstoneback/ScienceData/sw/ace/historic/epam/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace realtime sis
Current path :  /Users/russellstoneback/ScienceData/sw/ace/realtime/sis/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.

Working Instrument: sw ace historic sis
Current path :  /Users/russellstoneback/ScienceData/sw/ace/historic/sis/
Proposed path:  /Users/russellstoneback/ScienceData/sw/
No files found.
```

I've stubbed out support for creating file names on the fly as opposed to returning a pre-filled list. A `list_files` function can now return a dict formatted as 
```
{'format_str' : str, 'start_date': datetime, 'stop_date': datetime}
```                                   
and this will get pass to `instruments.methods.general.filename_creator` and set as a partial function in files. The filename_creator function still needs to be written. The current setup will work for loading by date, but iterating is a problem. See the comments in the code, but the short answer is either we need to create a generator function that pysat will use to make a list anyway (unless we update iteration again) or accept pre-filled lists.

Tasks remaining:
 - [x] Testing for directory update. We need to create a programmatic pysat install, with actual files.
 - [x] Apply changes to full on documentation
 - [x] Remove empty directories after transition
 - [x] Self-review on my part

rstoneback commented:
Content in #633 and #634 relevant here for further progress on translating files to new directory structure, as well as multiple directory support for loading files.

aburrell commented on pysat/utils/time.py:
```diff
@@ -6,7 +6,7 @@
 seasons, and calculation of solar local time
 """
-
+import datetime
```
Double import here, which doesn't conform to coding sty.e

aburrell reviewed (commented):
Found a small thing when checking out current status of branch.

rstoneback commented:
Thanks @aburrell for the merge.

rstoneback commented:
Friendly heads up I'll be playing some catch up over the weekend and Monday is meeting free and an actual pysat day.

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,206 @@
+        assert pysat.params['data_dirs'] == []
+
+        return
+
+
+
```
Many spaces!

rstoneback commented:
I'm starting to really regret expanding TravisCI tests! :P

rstoneback commented:
Ha! I finally got the test going. Note for future self. Using a `from pysat import params` in a pysat submodule does not work across a `reload(pysat)`. Switching submodules to `pysat.params` solves the issue. Similarly, we had a latent issue with `user_modules` in registry since it employed a `from pysat import user_modules`.

Next up I'm going to remove `pysat.user_modules` in favor of `pysat.params['user_modules']` since we now have two locations with the same information. There can only be one! That being said, I'm going to give myself and the website a few minutes to feel good about a green test status.

rstoneback commented:
I've run into an issue with the `update_data_directory_structure` method. For my first attempt I identified all the current files and tried to move them all to the new structure. This runs into issues with Instruments where list_files manages the actual files to create a list suitable for daily files. Thus the list of files we have doesn't match files on disk. We can move the files for Instruments where things are good but for irregular Instruments the plan would be to fail and tell users to move it themselves.

An alternative:
Rather than move files we can rename directories. The issue here is 
 - Non instrument files in those directories are now also in the renamed directory. If we go by file non Instrument files are left alone.
 - There can be a loss of information with this path. If the directory_template is switched to say {platform}/{name} and then users want to go back to {platform}/{name}/{tag}/{inst_id}, it won't work for Instruments with more than one tag or inst_id. The first combo of labels will rename the directory, and then when the next iteration of tag or inst_id combo Instrument is instantiated it won't find any files.

rstoneback commented:
I decided to stick with the move by file method.

rstoneback commented:
> I've run into an issue with the `update_data_directory_structure` method. 

I've sorted it out.

rstoneback commented on pysat/tests/test_constellation.py:
```diff
@@ -16,7 +16,8 @@ def setup(self):
         self.instruments = [pysat.Instrument('pysat', 'testing',
-                                             clean_level='clean')
+                                             clean_level='clean',
+                                             update_files=True)
```
We may need to set the temporary_file_list flag here. @jklenzing I believe I saw you had modified some testing configurations to prevent changes to a user setup. I was having problems with testing at one point and thus scattered `update_files=True` around. I may have reintroduced stuff you removed.

rstoneback commented:
I plan on tackling the .rst docs after there is a chance for feedback on the technical changes.

rstoneback commented:
I also think I stepped on a normalization of inputs pull that I missed the first time through. Open to suggestions on supporting user input plus Instrument defaults stored in pysat.params.

aburrell commented on docs/tutorial_parameters.rst:
```diff
@@ -0,0 +1,95 @@
+Parameters
+==========
```
```suggestion
pysat Parameters
================
```

aburrell commented on docs/tutorial_parameters.rst:
```diff
@@ -0,0 +1,95 @@
+   # Add new user parameter
+   pysat.params[user_param_str] = user_param_value
+
+All of the standard parameters above may be reset to defaults using
```
```suggestion
All of the standard parameters above may be restored to defaults using
```

aburrell commented on pysat/__init__.py:
```diff
@@ -47,83 +47,51 @@
-# get home directory
+# Get home directory
 home_dir = os.path.expanduser('~')
+# Set pysat directory path in home directory
+pysat_dir = os.path.join(home_dir, '.pysat')
 # Set directory for test data
```
```suggestion

# Set pysat directory path in home directory
pysat_dir = os.path.join(home_dir, '.pysat')

# Set directory for test data
```

aburrell commented on pysat/__init__.py:
```diff
@@ -47,83 +47,51 @@
     print(''.join(("\nHi there!  Pysat will nominally store data in the "
                    "'pysatData' directory at the user's home directory level. "
-                   "Run pysat.utils.set_data_dir to specify a different "
+                   "Assign `pysat.params['data_dirs']` a path to specify a "
```
```suggestion
                   "Set `pysat.params['data_dirs']` equal to a path that specifies a "
```

Possible better(?) wording??

aburrell commented on pysat/_files.py:
```diff
@@ -5,64 +5,93 @@
 
 import pandas as pds
-from pysat import data_dir as data_dir
+import pysat  # Needed to access pysat.params across reimports
+from pysat import logger, Instrument
```
Since we're importing pysat, recommend using `logger = pysat.logger` and `pysat.Instrument` below and removing second line.

aburrell commented on pysat/_files.py:
```diff
@@ -5,64 +5,93 @@
 
 class Files(object):
-    """Maintains collection of files for instrument object.
+    """Maintains collection of files plus associated methods.
```
```suggestion
    """Maintains collection of files and associated methods.
```

aburrell commented on pysat/_files.py:
```diff
@@ -106,116 +135,110 @@ class Files(object):
+                                         "Please set a top-level directory",
+                                         "path to store data using",
+                                         "`pysat.params['data_dirs'] = path`")))
+        # Get list of potential data directory paths from pysat. Construct
```
```suggestion

        # Get list of potential data directory paths from pysat. Construct
```

aburrell commented on pysat/_files.py:
```diff
@@ -309,16 +326,21 @@ def __getitem__(self, key):
+            full_fname = os.path.join(path, fname)
 
-            # Ensure the file exists and is a file
+            # Ensure it exists
```
```suggestion
            # Ensure the file exists
```

aburrell commented on pysat/_params.py:
```diff
@@ -0,0 +1,319 @@
+                                'file. Please check the locations, "./" or ',
+                                '"~/.pysat" for the file "pysat_settings.json"',
+                                '.'))
+                raise RuntimeError(estr)
```
Should this be OSError?

aburrell replied:
Still waiting for a response on this one.

rstoneback replied:
I missed that one somehow. I did change other errors in that file to OSError.

aburrell commented on pysat/_params.py:
```diff
@@ -0,0 +1,319 @@
+    def clear_and_restart(self):
+        """Clears all stored settings and sets pysat defaults
+
+        pysat parameters without a default value are set to [] """
```
```suggestion
        pysat parameters without a default value are set to [] 
        """
```

aburrell commented on pysat/_params.py:
```diff
@@ -0,0 +1,319 @@
+        return
+
+    def store(self):
+        """Store parameters to file.
+
+        Uses path in self.file_path.
```
```suggestion
        """Store parameters to file using the filename specified in self.file_path.
```

aburrell commented on pysat/instruments/methods/general.py:
```diff
@@ -208,3 +208,30 @@ def remove_leading_text(inst, target=None):
+
+    """
+
+    raise NotImplementedError()
```
Maybe add an error message?  Also add a Raises section to the docstring (here and elsewhere in this class)

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,227 @@
+# DOI:10.5281/zenodo.1199703
+# ----------------------------------------------------------------------------
+"""
+tests the pysat utils area
```
This seems wrong

aburrell replied:
And still needs to be addressed.

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,227 @@
+class TestBasics():
+    def setup(self):
+        """Runs before every method to create a clean testing setup."""
+        # store current pysat directory
```
Capitalize here and elsewhere

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,227 @@
+        """Runs before every method to create a clean testing setup."""
+        # store current pysat directory
+        self.stored_params = copy.deepcopy(pysat.params)
+        # set up default values
```
```suggestion

        # Set up default values
```

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,227 @@
+        pysat.params.store()
+        reload(pysat)
+
+    #######################
+    # test pysat data dir options
```
```suggestion
```

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,227 @@
+
+    #######################
+    # test pysat data dir options
+    def test_set_data_dirs_param_single(self):
+        """Update pysat directory via params, single string input"""
+        pysat.params['data_dirs'] = '.'
+        assert pysat.params['data_dirs'] == ['.']
+
+        # Check if next load of pysat remembers the change
+        reload(pysat)
+        assert pysat.params['data_dirs'] == ['.']
+
+    def test_set_data_dirs_param_with_list(self):
+        """Update pysat directories via pysat.params, list of strings"""
+        pysat.params['data_dirs'] = ['.', './']
+        assert pysat.params['data_dirs'] == ['.', './']
+
+        # Check if next load of pysat remembers the change
+        reload(pysat)
+        assert pysat.params['data_dirs'] == ['.', './']
```
These two can be parametrized

aburrell commented on pysat/tests/test_params.py:
```diff
@@ -0,0 +1,227 @@
+        reload(pysat)
+        assert pysat.params['data_dirs'] == ['.', './']
+
+    def test_set_data_dir_wrong_path(self):
+        """Update data_dirs with an invalid path form"""
+        with pytest.raises(ValueError):
+            pysat.params['data_dirs'] = 'not_a_directory'
+
+    def test_set_data_dir_bad_directory(self):
+        """Ensure you can't set data_dirs to bad path"""
+        with pytest.raises(ValueError) as excinfo:
+            pysat.params['data_dirs'] = '/fake/directory/path'
+        assert str(excinfo.value).find("don't lead to a valid") >= 0
```
These two can be parametrized

rstoneback commented:
I started planning/creating more test instruments to support the directory moving code but it will take a decent amount of time. On the other hand, we could use pysatSpaceWeather as a test package. I can register the instruments, download some data (are downloads supported on TravisCI for these?), then do the directory testing. Open to comments and suggestions!

aburrell commented:
That seems like a good plan to me!

rstoneback commented:
> That seems like a good plan to me!

Great! I checked on the download availability by looking at coveralls. There are definitely some ftp sources but there are enough non-ftp to work with to cover the full range of options.

rstoneback commented:
> But also don't forget about my last review.

:)

Sure won't. In this case your comments are the easy part. Working on hard -> easy.

rstoneback commented:
Finally! Tests are passing on TravisCI.

rstoneback commented:
Subpackages reviewed and corresponding pull requests:
 - [x] pysatSpaceWeather - https://github.com/pysat/pysatSpaceWeather/pull/42
 - [ ] pysatNASA
 - [ ] pysatModels
 - [ ] pysatIncubator
 - [ ] pysatMadrigal
 - [ ] pysatMissions
 - [ ] pysatSeasons

aburrell commented on pysat/__init__.py:
```diff
@@ -47,83 +47,53 @@
+        os.mkdir(pysat_dir)
+        os.mkdir(os.path.join(pysat_dir, 'instruments'))
+        os.mkdir(os.path.join(pysat_dir, 'instruments', 'archive'))
+        print('Created .pysat directory in home directory to store settings.')
```
Should this go to logging?  Most routines don't tell the user they're doing this.

aburrell commented on pysat/_files.py:
```diff
@@ -106,116 +136,111 @@ class Files(object):
+
+        # Ensure we have at least one path for pysat data directory
+        if len(pysat.params['data_dirs']) == 0:
+            raise RuntimeError(" ".join(("pysat's `data_dirs` hasn't been set.",
```
```suggestion
            raise NameError(" ".join(("pysat's `data_dirs` hasn't been set.",
```

aburrell commented on pysat/_files.py:
```diff
@@ -412,54 +450,70 @@ def _store(self):
         pandas.Series
-            Full path file names are indexed by datetime
-            Series is empty if there is no file list to load
+            file path names, indexed by datetime. Series is empty if no
```
```suggestion
            File path names, indexed by datetime. Series is empty if no
```

aburrell replied:
Also, it looks like the return variable now is `out`, so this needs to be updated further.

rstoneback replied:
I don't really understand the comment about `out`. There is a variable `out` in method below so I'll poke at that docstring.

aburrell replied:
Could be that.  From the display it appeared to be a part of this method.

aburrell replied:
Capitalization still stands, though.

rstoneback replied:
I accepted your suggestion.

Here's what my local system is showing,
```
        Returns
        -------
        pandas.Series
            File path names, indexed by datetime. Series is empty if no
            files are found.
```

aburrell commented on pysat/_instrument.py:
```diff
@@ -39,7 +40,8 @@ class Instrument(object):
     inst_id : string, optional
         identity within constellation
     clean_level : {'clean','dusty','dirty','none'}, optional
```
```suggestion
    clean_level : str or NoneType
```

Add string options to the description

Also, `, optional` isn't a typical feature of the docstrings.

rstoneback replied:
Accepted. That docstring has probably been around for awhile. Though we aren't really using it, the optional notation is part of the numpydoc standard.

```
If it is not necessary to specify a keyword argument, use optional:

x : int, optional
```

aburrell commented on pysat/utils/files.py:
```diff
@@ -395,3 +397,281 @@ def search_local_system_formatted_filename(data_path, search_str):
+def check_and_make_path(path):
+    """Checks if path exists, creates it if it doesn't.
+
+    Raises RuntimeError error if unable to complete.
```
This belongs in a `Raises` section.

aburrell commented on pysat/utils/files.py:
```diff
@@ -395,3 +397,281 @@ def search_local_system_formatted_filename(data_path, search_str):
+        Directory path without any file names. Creates all
+        necessary directories to complete the path.
+
+    Returns
+    -------
+    None
```
Remove if nothing is returned.

aburrell reviewed (commented):
Found some areas that weren't addressed from my last review and also new things to fix.

rstoneback commented:
@aburrell and @jklenzing I've uploaded a short tutorial section on moving from pysat v2.x to v3.0.

aburrell reviewed (approved):
I still approve!

rstoneback commented:
> I still approve!

Wonderful! Thanks again for reviewing this large pull. I pushed one last minor change and now I'm done.

I realized after the last push, but before your feedback, that the transition tutorial needed a quick note about the registry. I added that in plus I added a few sentences over in the registry class to help point users to registering instruments by module.
